### PR TITLE
refactor: [M3-8203] - Refactor Region Select

### PR DESCRIPTION
**Pull request by bnussman-akamai**
## Why ❓ 
- When I started working on addressing the Linode Create v2 side effect renders, I was unable to achieve what I wanted to easily with the current implementation. It wasn't easy to get the full `Region` object in the `onChange` handler, which I need because I want to reference the region capabilities in my`onChange`. I'm doing this PR first so that I can more easily get the raw `Region` value onChange, also allowing us to clean up the `RegionSelectOption` abstraction.

## Changes 🔁
- Makes RegionSelect and RegionMultiSelect use `Region` type instead of the `RegionSelectOption` abstraction
- Changes prop `handleSelection` to `onChange` to be more consistent with most selects
- Changes prop `selectedId` to `value` to be more consistent with most selects
- Changes prop `handleDiabledRegion` to `disabledRegions` to simplify disabling region options
- Removes custom `isClearable` prop in favor of MUI's `disableClearable` so we can retain the typesafety of `onChange` being `(region:Region | null) => void` when `disableClearable` is `false` and  `(region: Region) => void` when `disableClearable` is `true`

## Preview 📷

> [!note]
> No UI changes expected

## How to test 🧪

- Test the Region Select throughout the application
- Test the RegionMultiSelect, which can be found on the Object Storage Access Keys page with **Mutli-Cluster** turned on
- Make sure Account Availability is working as expected
- Verify regions are shown in the correct order (should be the same order as production)
- Verify Distributed Compute Region icons render when expected

## As an Author I have considered 🤔

- [ ] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx [hunk 1]
```diff
@@ -134,20 +107,19 @@ export const RegionSelect = React.memo((props: RegionSelectProps) => {
         textFieldProps={{
           ...props.textFieldProps,
           InputProps: {
-            endAdornment: regionFilter !== 'core' &&
```
Comment (bnussman-akamai) [inline packages/manager/src/components/RegionSelect/RegionSelect.tsx:137]: @hana-linode 

Can I safely remove the `regionFilter !== 'core'` condition? I can't quite tell what that is doing

File: packages/manager/src/components/RegionSelect/RegionSelect.types.ts [hunk 2]
```diff
@@ -37,71 +35,50 @@ export interface RegionSelectProps
...
    */
   currentCapability: Capabilities | undefined;
-  handleDisabledRegion?: (
-    region: Region
-  ) => RegionSelectOption['disabledProps'];
-  handleSelection: (id: string) => void;
+  /**
+   * A key/value object for disabling regions by their ID.
+   */
+  disabledRegions?: Record<string, DisableRegionOption>;
```
Comment (bnussman-akamai) [inline packages/manager/src/components/RegionSelect/RegionSelect.types.ts:41]: @abailly-akamai Do you have any opinions here? A key value object feels a bit more intuitive to me, but I'd also be fine to revert this to a function style if you think it's better

Comment (abailly-akamai) [reply packages/manager/src/components/RegionSelect/RegionSelect.types.ts:41]: This is a big PR with a lot of changes affecting a lot of features. I'll try and take some time tomorrow to review it but it isn't super high priority right now

Comment (abailly-akamai) [reply packages/manager/src/components/RegionSelect/RegionSelect.types.ts:41]: This is fine by me. I don't have a strong opinion for one or the other.

**Review by abailly-akamai — COMMENTED** [review]
Refactor looks good overall and is a welcome improvement 👍 

A few things:
- Storybook `RegionSelect` has a regression. When selecting a value and bluring out the initial value gets reset
- Storybook `RegionMultiSelect` is plain broken

**Comment by bnussman-akamai** [comment]
Thank you for that feedback @abailly-akamai  🙏 

I pushed some changes to fix up the Storybook stories (f6d211d) and re-enabled the unit tests for the multi-select (25631ac)

File: packages/manager/src/components/RegionSelect/RegionMultiSelect.test.tsx [hunk 3]
```diff
@@ -55,11 +54,12 @@ describe('RegionMultiSelect', () => {
     screen.getByRole('combobox', { name: 'Regions' });
   });
 
-  it('should be able to select all the regions correctly', () => {
+  // Re-enable if we allow RegionMultiSelect to be uncontrolled
+  it.skip('should be able to select all the regions correctly', () => {
```
[the remark this answers is not in the document]
Comment (bnussman-akamai) [reply packages/manager/src/components/RegionSelect/RegionMultiSelect.test.tsx]: Added back these tests in 25631ac with some minor modifications to account for the fact the component is intended to be controlled

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx (hunk 1 above)
Comment (hana-linode) [reply packages/manager/src/components/RegionSelect/RegionSelect.tsx:137]: @bnussman I think it was to check for edge regions but looks like we're checking the site type too. I'll say we can tentatively remove it. The adornment won't be around after Gecko Beta anyway

Comment (bnussman-akamai) [reply packages/manager/src/components/RegionSelect/RegionSelect.tsx:137]: Sounds good! I'll keep an eye out for any weird behavior regarding when the icon renders 👁️

**Review by abailly-akamai — APPROVED** [review]
Great refactor! I tested:
- RegionSelect and RegionMultiSelect behavior in various flows ✅ 
- Disabled Options logic & rendering ✅
- Storybook ✅

And did not find regressions other than the ones already addressed in the feedback.

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx [hunk 4]
```diff
@@ -156,18 +133,21 @@ export const RegionSelect = React.memo((props: RegionSelectProps) => {
...
         label={label ?? 'Region'}
         loading={accountAvailabilityLoading}
         loadingText="Loading regions..."
         noOptionsText="No results"
-        options={options}
+        onChange={onChange}
+        options={regionOptions}
         placeholder="Select a Region"
-        value={selectedRegion}
+        value={selectedRegion as Region}
```
Comment (hkhalil-akamai) [inline packages/manager/src/components/RegionSelect/RegionSelect.tsx:150]: Is `selectedRegion` guaranteed to be non-null? If so, is there a way we can remove this casting?

File: packages/manager/src/components/RegionSelect/RegionSelect.utils.tsx [hunk 5]
```diff
@@ -1,114 +1,72 @@
...
 const NORTH_AMERICA = CONTINENT_CODE_TO_CONTINENT.NA;
 
-/**
- * Returns an array of OptionType objects for use in the RegionSelect component.
- * Handles the disabled state of each region based on the user's account availability or an optional custom handler.
- * Regions are sorted alphabetically by region, with North America first.
- *
- * @returns An array of RegionSelectOption objects
- */
+interface RegionSelectOptionsOptions {
```
Comment (hkhalil-akamai) [inline packages/manager/src/components/RegionSelect/RegionSelect.utils.tsx:18]: First I thought this name was a typo but it does make sense it's just funny 😹

File: packages/manager/src/components/RegionSelect/RegionSelect.utils.tsx [hunk 6]
```diff
@@ -1,114 +1,72 @@
...
+        currentCapability &&
+        !region.capabilities.includes(currentCapability)
+      ) {
+        return false;
+      }
+      if (regionFilter && region.site_type !== regionFilter) {
+        return false;
+      }
+      return true;
     })
```
Comment (hkhalil-akamai) [inline packages/manager/src/components/RegionSelect/RegionSelect.utils.tsx:41]: This is equivalent:
```suggestion
    .filter(
      (region) =>
        (!currentCapability ||
          region.capabilities.includes(currentCapability)) &&
        (!regionFilter || region.site_type == regionFilter)
    )
```

File: packages/manager/src/components/RemovableSelectionsList/RemovableSelectionsList.tsx [hunk 7]
```diff
@@ -117,9 +117,9 @@ export const RemovableSelectionsList = (
   // used to determine when to display a box-shadow to indicate scrollability
   const listRef = React.useRef<HTMLUListElement>(null);
   const [listHeight, setListHeight] = React.useState<number>(0);
-  const [removingItemId, setRemovingItemId] = React.useState<null | number>(
-    null
-  );
+  const [removingItemId, setRemovingItemId] = React.useState<
```
[the remark this answers is not in the document]
Comment (hkhalil-akamai) [reply packages/manager/src/components/RemovableSelectionsList/RemovableSelectionsList.tsx:120]: imo we should stick with `string` if that's the type on `Region` but I'm fine with having as a follow-up change.

File: packages/manager/src/features/ObjectStorage/AccessKeyLanding/AccessKeyRegions/AccessKeyRegions.tsx [hunk 8]
```diff
@@ -15,7 +15,7 @@ interface Props {
   selectedRegion: string[];
 }
 
-const sortRegionOptions = (a: RegionSelectOption, b: RegionSelectOption) => {
+const sortRegionOptions = (a: Region, b: Region) => {
   return sortByString(a.label, b.label, 'asc');
 };
```
Comment (hkhalil-akamai) [inline packages/manager/src/features/ObjectStorage/AccessKeyLanding/AccessKeyRegions/AccessKeyRegions.tsx:20]:
```suggestion
const sortRegionOptions = (a: Region, b: Region) => sortByString(a.label, b.label, 'asc');
```

File: packages/manager/src/features/ObjectStorage/AccessKeyLanding/AccessKeyRegions/SelectedRegionsList.tsx [hunk 9]
```diff
@@ -29,9 +29,9 @@ const SelectedRegion = ({ selection }: LabelComponentProps) => {
       }}
     >
       <StyledFlagContainer>
-        <Flag country={selection.data.country} />
+        <Flag country={selection.country} />
       </StyledFlagContainer>
-      {selection.label}
+      {selection.label} ({selection.id})
```
Comment (hkhalil-akamai) [inline packages/manager/src/features/ObjectStorage/AccessKeyLanding/AccessKeyRegions/SelectedRegionsList.tsx:34]: This looks like a user-facing change -- intentional?

File: packages/manager/src/components/RemovableSelectionsList/RemovableSelectionsList.tsx [hunk 10]
```diff
@@ -22,7 +22,7 @@ export type RemovableItem = {
   // as 'any' because we do not know what types they could be.
   // Trying to type them as 'unknown' led to type errors.
   [key: string]: any;
-  id: number;
+  id: number | string;
```
Comment (hkhalil-akamai) [inline packages/manager/src/components/RemovableSelectionsList/RemovableSelectionsList.tsx:25]: I don't love the type-safety on this interface. In a follow-up, can we use generics to remove the `any` (and possibly also narrow the type of `id`)?

Comment (hkhalil-akamai) [reply packages/manager/src/components/RemovableSelectionsList/RemovableSelectionsList.tsx:25]: e.g.,

This type could be rewritten as:
```
export type RemovableItem<T, Id extends number | string = number | string> = {
  id: Id;
  label: string;
} & T;
```

Then in SelectedRegionsList.tsx:
```
interface LabelComponentProps {
  selection: RemovableItem<Region>;
}
```

File: packages/manager/src/features/ObjectStorage/AccessKeyLanding/AccessKeyRegions/SelectedRegionsList.tsx [hunk 11]
```diff
@@ -41,14 +41,12 @@ export const SelectedRegionsList = ({
   selectedRegions,
 }: SelectedRegionsProps) => {
   const handleRemove = (item: RemovableItem) => {
-    onRemove(item.value);
+    onRemove(item.id as string);
```
Comment (hkhalil-akamai) [inline packages/manager/src/features/ObjectStorage/AccessKeyLanding/AccessKeyRegions/SelectedRegionsList.tsx:44]: With some improvements to the `RemovableItem` type, we can get rid of this casting. For a follow-up

**Review by hkhalil-akamai — APPROVED** [review]
Approved with some minor suggestions and a possible refactor in a follow-up change. Also don't forget to add changeset(s).

Great work, this is a big improvement on `RegionSelect`.

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx (hunk 4 above)
Comment (bnussman-akamai) [reply packages/manager/src/components/RegionSelect/RegionSelect.tsx:150]: I couldn't find a good way to fix the casting

File: packages/manager/src/features/ObjectStorage/AccessKeyLanding/AccessKeyRegions/SelectedRegionsList.tsx (hunk 9 above)
Comment (bnussman-akamai) [reply packages/manager/src/features/ObjectStorage/AccessKeyLanding/AccessKeyRegions/SelectedRegionsList.tsx:34]: This does not result in any UI change.

Previously `selection.label` wasn't the raw region label. It was the region label with the ID appended. This is one of the big reasons I did this refactor 😅 This is a lot more explicit and true to the `Region` type

File: packages/manager/src/components/RemovableSelectionsList/RemovableSelectionsList.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (bnussman-akamai) [reply packages/manager/src/components/RemovableSelectionsList/RemovableSelectionsList.tsx:120]: I believe VPC uses this and needs `number` so it needs to be both for now until we can make this generic

File: packages/manager/src/components/RegionSelect/RegionSelect.utils.tsx (hunk 6 above)
Comment (abailly-akamai) [reply packages/manager/src/components/RegionSelect/RegionSelect.utils.tsx:41]: This is far less readable tho. I personally prefer the original suggestion

**Comment by hkhalil-akamai** [comment]
Saw this was merged but it seems like no changeset was included. Not sure if that was intentional.

**Pull request merged**